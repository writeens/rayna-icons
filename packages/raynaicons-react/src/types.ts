import { RefAttributes, ForwardRefExoticComponent, SVGProps } from "react";

export type RaynaProps = RefAttributes<SVGSVGElement> &
  Partial<SVGProps<SVGSVGElement>> & {
    title?: string;
    size?: string | number;
  };

export type RaynaIcon = ForwardRefExoticComponent<RaynaProps>;
