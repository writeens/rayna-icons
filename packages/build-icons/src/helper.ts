/** Convert icon name (abc-def-ghi) to pascal case (AbcDefGhi) */
export const iconNameToPascalCase = (name: string) => {
  const parts = name.split("-");
  return parts
    .map((part) => `${part[0].toUpperCase()}${part.slice(1).toLowerCase()}`)
    .join("");
};
