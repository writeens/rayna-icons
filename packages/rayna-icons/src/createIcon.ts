import { IconNode, RaynaIcon, RaynaIconAttributes } from "./types";

const toSVG = (icon: IconNode) => (attributes?: RaynaIconAttributes) => {
  const {
    width: defaultWidth,
    height: defaultHeight,
    viewBox: defaultViewBox,
  } = icon;
  const defaultClassName = `rayna rayna-${icon.name}`;

  const newAttributes: RaynaIconAttributes = {
    ...attributes,
    width: attributes && attributes.width ? attributes.width : defaultWidth,
    height: attributes && attributes.height ? attributes.height : defaultHeight,
    viewBox:
      attributes && attributes.viewBox ? attributes.viewBox : defaultViewBox,
    class:
      attributes && attributes.class
        ? `${defaultClassName} ${attributes.class}`
        : defaultClassName,
  };

  const attributesAsString = Object.keys(newAttributes)
    .map((key) => `${key}="${newAttributes[key]}"`)
    .join(" ");

  return `<svg ${attributesAsString}>${icon.path}</svg>`;
};

export const createIcon = (icon: IconNode): RaynaIcon => ({
  name: icon.componentName,
  width: icon.width,
  height: icon.height,
  viewBox: icon.viewBox,
  path: icon.path,
  keywords: icon.keywords,
  toSVG: toSVG(icon),
});
