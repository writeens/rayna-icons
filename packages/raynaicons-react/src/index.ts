/* Types exports */
export * from "./types";

/* Icons exports */
export * from "./__generated__";
