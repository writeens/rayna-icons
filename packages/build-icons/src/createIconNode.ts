/* eslint-disable import/no-relative-packages */
import { load } from "cheerio";
import { parseSync } from "svgson";
import keywordJSON from "../../../keywords.json" assert { type: "json" };
import { iconNameToPascalCase } from "./helper";
import { optimizeSVG } from "./optimizeSVG";
import { IconNode } from ".";

const keywords = keywordJSON as Record<string, string[]>;

export const createIconNode = (
  svg: ReturnType<typeof optimizeSVG>[number],
): IconNode => {
  const SVGElement = load(svg.content)("svg");
  const SVGWidth = parseInt(SVGElement.attr("width") || "", 10);
  const SVGHeight = parseInt(SVGElement.attr("height") || "", 10);
  const SVGViewBox = SVGElement.attr("viewBox") || "";
  const SVGPath = SVGElement.html() ?? "";
  const SVGName = `${svg.fileName}-${svg.type}`;
  const ast = parseSync(svg.content);

  if (!SVGHeight) {
    throw new Error(`${SVGName}: Missing height attribute.`);
  }

  if (!SVGWidth) {
    throw new Error(`${SVGName}: Missing width attribute.`);
  }

  if (!SVGViewBox) {
    throw new Error(`${SVGName}: Missing viewBox attribute.`);
  }

  const viewBoxPattern = /0 0 ([0-9]+) ([0-9]+)/;

  const viewBoxMatches = SVGViewBox.match(viewBoxPattern);

  if (!viewBoxMatches) {
    throw new Error(
      `${SVGName}: Invalid viewBox attribute. Ensure the viewBox is in the following format: "0 0 <width> <height>"`,
    );
  }

  const [, viewBoxWidth, viewBoxHeight] = viewBoxMatches;

  if (SVGWidth !== parseInt(viewBoxWidth, 10)) {
    throw new Error(
      `${SVGName}: Invalid width attribute. The width attribute and the viewBox width do not match`,
    );
  }

  if (SVGHeight !== parseInt(viewBoxHeight, 10)) {
    throw new Error(
      `${SVGName}: Invalid height attribute. The height attribute and the viewBox height do not match`,
    );
  }

  return {
    name: SVGName,
    componentName: iconNameToPascalCase(SVGName),
    width: SVGWidth,
    height: SVGHeight,
    viewBox: SVGViewBox,
    keywords: keywords[svg.fileName],
    path: SVGPath,
    ast: {
      children: ast.children.map((node) => ({
        name: node.name,
        attributes: node.attributes,
      })),
    },
  };
};
