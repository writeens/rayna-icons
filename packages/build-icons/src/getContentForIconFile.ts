import prettier from "prettier";
import { IconNode } from ".";

export const GENERATED_HEADER = "/* THIS FILE IS GENERATED. DO NOT EDIT IT. */";

/** Generate the content of an icon file from an icon node */
export const getContentForIconFile = async (icon: IconNode) => {
  const code = `
        ${GENERATED_HEADER}

        import {createIcon} from '../createIcon'
        import type {RaynaIcon} from '../types'

        /**
         * @name ${icon.componentName}
         * @description Rayna SVG icon node.
         * 
         * * */

        const ${icon.componentName}: RaynaIcon = createIcon(${JSON.stringify(icon)})

        export default ${icon.componentName}
        `;

  const formattedCode = await prettier.format(code, {
    semi: true,
    printWidth: 80,
    parser: "babel-ts",
  });

  return formattedCode;
};
