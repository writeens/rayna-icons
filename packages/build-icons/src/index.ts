import path from "path";
import { appendFile, outputFile, outputFileSync } from "fs-extra";
import { INode } from "svgson";
import { createIconNode } from "./createIconNode";
import { getContentForIconExport } from "./getContentForIconExport";
import {
  GENERATED_HEADER,
  getContentForIconFile,
} from "./getContentForIconFile";
import { optimizeSVG } from "./optimizeSVG";

/** Build icon files from icons and append their exports to a barrel file */
export const buildIconFiles = ({
  outputDirectory,
  rawIconDirectory,
  // eslint-disable-next-line no-use-before-define
}: BuildOptions) => {
  const exportsFileLocation = path.join(outputDirectory, "index.ts");

  const filePattern = path.join(rawIconDirectory, "*", "*.svg");

  /** Optimize SVGs */
  const optmizedSVGData = optimizeSVG(filePattern);

  /** Create icon node */
  const iconNodes = optmizedSVGData.map(createIconNode);

  /** Create the barrel file with a header */
  outputFileSync(exportsFileLocation, `${GENERATED_HEADER}\n`);

  Promise.all(
    iconNodes.map(async (iconNode) => {
      const outputFileLocation = path.join(
        outputDirectory,
        `${iconNode.name}.ts`,
      );

      /** Create the icon file */
      getContentForIconFile(iconNode).then((contentForIconFile) => {
        outputFile(outputFileLocation, contentForIconFile);
      });

      /** Add the new icon file as an entry in index.ts (barrel file) */
      getContentForIconExport(iconNode).then((contentForIconExport) => {
        appendFile(exportsFileLocation, contentForIconExport);
      });
    }),
  )
    .then(() => {
      console.log(
        "Successfully generated the icon file and exports for",
        iconNodes.length,
        "icons",
      );
    })
    .catch((error) => {
      console.error(error);
    });
};

export type IconNode = {
  name: string;
  componentName: string;
  width: number;
  height: number;
  viewBox: string;
  keywords: string[];
  path: string;
  ast: {
    children: Pick<INode, "name" | "attributes">[];
  };
};

export type BuildOptions = {
  outputDirectory: string;
  rawIconDirectory: string;
};
