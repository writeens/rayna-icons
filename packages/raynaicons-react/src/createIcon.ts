import { IconNode } from "@writeens/build-icons";
import { forwardRef, createElement } from "react";
import type { RaynaIcon, RaynaProps } from "./types";

export const createIcon = (iconNode: IconNode): RaynaIcon => {
  const Component = forwardRef<SVGSVGElement, RaynaProps>(
    ({ width, height, title, size = 24, className, ...rest }, ref) =>
      createElement(
        "svg",
        {
          ref,
          width: width ?? size ?? iconNode.width,
          height: height ?? size ?? iconNode.height,
          className: ["rayna", `rayna-${iconNode.name}`, className].join(" "),
          ...rest,
        },
        [
          title && createElement("title", {}, title),
          iconNode.ast.children.map((node) =>
            createElement(node.name, node.attributes),
          ),
        ].filter(Boolean),
      ),
  );

  Component.displayName = `${iconNode.componentName}`;

  return Component;
};
