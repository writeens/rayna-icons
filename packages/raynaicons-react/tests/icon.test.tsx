import { render, screen } from "@testing-library/react";
import { it, describe, expect } from "vitest";
// Build the iconfiles first to remove squiggly lines
import { FirstAidFilled } from "../src/__generated__";

describe("FirstAidFilled", () => {
  it("is a valid react component", () => {

    const { container } = render(<FirstAidFilled />)

    expect(container).toMatchSnapshot()
  });

  it("should pass options correctly to the SVG", () => {

    const testId = "icon"

    const { getByTestId } = render(<FirstAidFilled data-testid={testId} size={40} className="custom-class" />)

    const svg = getByTestId(testId)

    screen.debug(svg)

    expect(svg).toHaveAttribute("class", "rayna rayna-first-aid-filled custom-class")
    expect(svg).toHaveAttribute("width", "40")
    expect(svg).toHaveAttribute("height", "40")
  })
});
