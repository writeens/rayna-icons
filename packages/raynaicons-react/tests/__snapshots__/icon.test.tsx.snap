// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FirstAidFilled > is a valid react component 1`] = `
<div>
  <svg
    class="rayna rayna-first-aid-filled"
    height="24"
    width="24"
  >
    <path
      clip-rule="evenodd"
      d="M8.038 6.667H5.722A2.22 2.22 0 0 0 3.5 8.889v.482l7.266 2.698c.412.153.848.231 1.287.231h.894c.44 0 .875-.078 1.287-.231L21.5 9.37V8.89a2.22 2.22 0 0 0-2.222-2.222h-2.316A3.67 3.67 0 0 0 13.433 4h-1.866a3.67 3.67 0 0 0-3.529 2.667m2.195 0h4.534A1.66 1.66 0 0 0 13.433 6h-1.866c-.546 0-1.03.262-1.334.667M13.5 8.75a1 1 0 1 0-2 0 1 1 0 1 0 0 2 1 1 0 1 0 2 0 1 1 0 1 0 0-2"
      fill="#101928"
      fill-rule="evenodd"
    />
    <path
      d="m5.027 17.302-.627-6.19 6.696 2.479a3 3 0 0 0 1.042.187h.724a3 3 0 0 0 1.042-.187l6.696-2.48-.627 6.191A3 3 0 0 1 16.988 20H8.012a3 3 0 0 1-2.985-2.698"
      fill="#101928"
    />
  </svg>
</div>
`;
